### PR TITLE
Refactor searchpromotions index view to CBV

### PR DESCRIPTION
**Pull request by rohitsrma**
In this PR I've converted search promotions index view from FBV to CBV as a part of #8365.

### CODE REVIEW

**Comment by rohitsrma** [comment]
@lb- I'm uncertain whether my approach is correct or not. Could you please provide your insights on this?

**Review by laymonage — CHANGES_REQUESTED** [review]
Hey, thanks for giving this a go! This is a great start. Could you look into how `generic.IndexView` (and potentially its superclass, `BaseListingView`) works, and make it so that there's as little overrides as possible? This would allow us to use the new designs (Universal Listings) much easier.

For reference, see each commit in #11331 and #11304 on how I refactored the forms and documents views (although the latter still uses the old designs). Thanks again!

**Comment by rohitsrma** [comment]
Thanks @laymonage, for your valuable feedback and suggestions. I've followed your recommendations, and everything is functioning correctly except for the **`search`** functionality. I'm encountering the same bug described in https://github.com/wagtail/wagtail/pull/11035#issuecomment-1807109543 i.e,when entering text into the search field the UI become distorted.

### Screenshot of Bug:

image

File: wagtail/contrib/search_promotions/admin_urls.py
Comment (laymonage) [inline wagtail/contrib/search_promotions/admin_urls.py:7]: @rohitsrma add another view for the results here, i.e.

```suggestion
    path("results/", views.IndexView.as_view(results_only=True), name="index_results"),
```

File: wagtail/contrib/search_promotions/views.py
Comment (laymonage) [inline wagtail/contrib/search_promotions/views.py:8]: Be careful now that we have translatable strings at the class definition level (not inside a function). These strings will need to use `gettext_lazy`, so that they're not immediately translated at application startup.

File: wagtail/contrib/search_promotions/views.py
Comment (laymonage) [inline wagtail/contrib/search_promotions/views.py:30]: Then add `index_results_url_name` with the new URL, i.e.

```suggestion
    index_url_name = "wagtailsearchpromotions:index"
    index_results_url_name = "wagtailsearchpromotions:index_results"
```

File: wagtail/contrib/search_promotions/views.py
Comment (laymonage) [inline wagtail/contrib/search_promotions/views.py:60]: We might be able to define `context_object_name = "queries"` on the view class and remove this altogether? Otherwise we'll be calling `get_queryset()` twice.

**Review by laymonage — COMMENTED** [review]
Thanks for continuing to work on this! The results fragment bug is a result of us not having the separate results view set up. I've left comments on how to fix it, as well as a few others around the need for `gettext_lazy`.

File: wagtail/contrib/search_promotions/views.py
Comment (laymonage) [inline wagtail/contrib/search_promotions/views.py:26]:
```suggestion
    page_title = gettext_lazy("Search Terms")
```

(and other translatable strings at class definition level)

**Comment by rohitsrma** [comment]
Thanks @laymonage, I followed your suggestions but the bug still persist!

File: wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/index.html
Comment (laymonage) [inline wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/index.html:5]: We need to also change the `search_results_url` here, although ideally we should try whether we can completely remove this block override (or override sub-blocks in `generic/index.html`) instead. Then, setting `_show_breadcrumbs = True` on the view class would allow this view to use the new designs.

Comment (rohitsrma) [reply wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/index.html:5]: @laymonage Sorry i forgot to update this URL :)

I attempted removing the block override from `wagtailsearchpromotions/index.py`, and here's the updated UI:

image

Is it fine?

**Comment by rohitsrma** [comment]
Now everything is working well!

**Comment by rohitsrma** [comment]
@laymonage Currently the issue is there are no hyperlinks to promoted results pages. I have tried to address this by overriding the `Column` class to add links to the page edit view. However, one of the tests is still failing. Here is what I have done:

```
class SearchPromotionColumn(Column):
    edit_url_name = "wagtailadmin_pages:edit"
    cell_template_name = "wagtailsearchpromotions/search_promotion_column.html"

    def get_url(self, instance):
        query_instance = SearchPromotion.objects.filter(query=instance).latest('id')
        if query_instance.page and query_instance.page.pk:
            return reverse(self.edit_url_name, args=(quote(query_instance.page.pk),))
        return query_instance.external_link_url if query_instance else ""

    def get_cell_context_data(self, instance, parent_context):
        context = super().get_cell_context_data(instance, parent_context)
        search_promotion = SearchPromotion.objects.filter(query=instance).latest('id')
        context["url"] = self.get_url(instance)
        context["external_link_text"] = search_promotion.external_link_text
        return context
```

I know my approach is very complicated and not the optimal one, but it is doing the job. I would greatly appreciate your suggestions on this!"

File: wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/list.html
Comment (laymonage) [inline wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/list.html:46]: As we can see here, we're displaying all the editor picks for this query object in a single row. We need to extract this template into a custom `Column` subclass and use that in the view's `columns`.

For an example, see the `ReferencesColumn` implementation for the usage view:

https://github.com/wagtail/wagtail/blob/e418cd6e5b8e0f115462b8298a8a630154754231/wagtail/admin/ui/tables/__init__.py#L358-L384

https://github.com/wagtail/wagtail/blob/e418cd6e5b8e0f115462b8298a8a630154754231/wagtail/admin/views/generic/usage.py#L109-L117

In this case, this column is specific to the `search_promotions` contrib app, so just put it in the same `views.py` above the `IndexView`.

File: wagtail/contrib/search_promotions/views.py
Comment (laymonage) [inline wagtail/contrib/search_promotions/views.py:26]: While we're at this, can we rename the template to `index_results.html` for consistency with the generic `IndexView`?

```suggestion
    results_template_name = "wagtailsearchpromotions/index_results.html"
```

(Rename the file too)

File: wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/list.html
Comment (laymonage) [inline wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/list.html:2]: This file is now unused, isn't it? Let's just remove it if that's the case.

**Review by laymonage — CHANGES_REQUESTED** [review]

**Comment by laymonage** [comment]
@rohitsrma thanks! You're correct that we need to create a custom `Column` subclass. However, let's not do any queries inside it. Instead, use `prefetch_related` from the `Column` subclass and try to reuse what we had in the old templates.

File: wagtail/contrib/search_promotions/views.py
Comment (rohitsrma) [reply wagtail/contrib/search_promotions/views.py:66]: @laymonage Sorry i am little confused!

```
def get_base_queryset(self):
        queryset = self.model.objects.filter(editors_picks__isnull=False).distinct()
        queryset = queryset.annotate(views=functions.Coalesce(Sum("daily_hits__hits"), 0))
        queryset = queryset.prefetch_related("editors_picks", "editors_picks__page")
        return queryset
```
When using `prefetch_related`, what should I use as the name in `SearchPromotionColumn`? When trying `editors_picks__page__title`, I received an error message:

_`editors_picks__page__title' does not resolve to an item that supports prefetching - this is an invalid parameter to prefetch_related().`_

So, how can i access page titles?

Comment (laymonage) [reply wagtail/contrib/search_promotions/views.py:66]: The two aren't directly related. You specify `editors_picks` and `editors_picks__page` in `prefetch_related` so that the instances of those models are queried in bulk. Then, you can pass `editors_picks` to the column. This means the `value` variable in the `SearchPromotionColumn` would be the `editors_picks`, i.e. a set of `SearchPromotion` instances. From there you can loop the `value` to get each `SearchPromotion` instance, so essentially in the templates you "just" replace `{% for searchpick in query.editors_picks.all %}` with `{% for searchpick in value %}`.

Comment (rohitsrma) [reply wagtail/contrib/search_promotions/views.py:66]: @laymonage Thanks a lot!

I properly understood what you said. But, I am just getting a little issue.





> Then, you can pass `editors_picks` to the column.

As you told to pass the `editors_picks` to the column. But, I think we cannot access the related values (like - Foreignkeys) directly, isn't it?

### Column:
```
SearchPromotionColumn(
            "editors_picks",
            label=_("Promote"),
        ), 

```

Therefore, I am getting an error that says:

`create_reverse_many_to_one_manager.<locals>.RelatedManager.__call__() missing 1 required keyword-only argument: 'manager'`

So, how can i approach this?

Comment (laymonage) [reply wagtail/contrib/search_promotions/views.py:66]: Does `editors_picks.all` work?

Comment (rohitsrma) [reply wagtail/contrib/search_promotions/views.py:66]: @laymonage I tried but still getting the same error.

Comment (rohitsrma) [reply wagtail/contrib/search_promotions/views.py:66]: Thankyou, finally it worked 🙌!

**Comment by rohitsrma** [comment]
Thanks @laymonage, now the PR is ready for review!

File: wagtail/contrib/search_promotions/views.py
Comment (laymonage) [inline wagtail/contrib/search_promotions/views.py:67]: Unfortunately there seems to be an issue here, the daily hits will be incorrect if the query has more than one `SearchPromotion`. The `editors_picks__isnull=False` causes a `JOIN` with the `SearchPromotion` table, and the `SUM` is done before the `DISTINCT`, so there will be multiple rows for the same `Query` row with different `SearchPromotion` rows. As a result, the `SUM` will be multiplied by the number of `SearchPromotion`s that the `Query` has.

To verify this, you can try using bakerydemo and adding another promoted result for the `patisserie` query. With each promoted search result added, the views count will be multiplied by the amount of promoted search results. For example, with one result it has 15 views, and if you add another result it will become 30. This wasn't an issue with the original code because it uses the `Query.hits` property, but it does a query for each item in the results (N+1 query problem).

One way to solve this that I found is by using a subquery for doing the `editors_picks__isnull=False` filter, e.g.

```suggestion
        has_promotions = SearchPromotion.objects.values_list("query_id", flat=True)
        queryset = self.model.objects.filter(pk__in=has_promotions)
```

Could you add a test case for this too, i.e. creating a `Query` with multiple `SearchPromotion`s, and ensuring the `views` column has the correct value? If not, I can try adding it as I merge.

File: wagtail/contrib/search_promotions/views.py
Comment (laymonage) [inline wagtail/contrib/search_promotions/views.py:36]: I was a bit surprised that there were no failing test cases when we didn't have this, so we likely need some tests to ensure that users without `add`/`change`/`delete` permissions for the `SearchPromotion` model should not be able to access the view.

Let me know if you'd like to try adding the test, otherwise I can do it as I merge 😃

**Review by laymonage — CHANGES_REQUESTED** [review]
Thanks, this is really great work so far! I was *this* close to merging it, but I found an issue with the views count. I've found the fix though, but this will need some more tests. If you're still up for working on this, you can try adding those tests I mentioned, but otherwise I can add them as I merge (maybe tomorrow or the day after). Thanks again!

File: wagtail/contrib/search_promotions/views.py
Comment (Stormheg) [reply wagtail/contrib/search_promotions/views.py:67]: Chipping in to say Sage and I had a lot of fun discussing this problem and its solutions over a call! Nice brain teaser🧠

**Comment by lb-** [comment]
Awesome to see the collaboration here, thanks @rohitsrma for keeping at it and Sage/Storm for bringing your knowledge and skills to the reviews.

File: wagtail/contrib/search_promotions/views.py
Comment (rohitsrma) [reply wagtail/contrib/search_promotions/views.py:67]: @Stormheg glad to hear it!✌️

File: wagtail/contrib/search_promotions/views.py
Comment (rohitsrma) [reply wagtail/contrib/search_promotions/views.py:36]: @laymonage I've added the test for permissions. please let me know if something is needed in it. I'm also trying to add tests for the views column.

**Comment by rohitsrma** [comment]
@laymonage Thanks a lot, I've made the updates. Please let me know if any changes are required!

File: wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/search_promotion_column.html
Comment (laymonage) [inline wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/search_promotion_column.html:15]: Needs a newline at end of file

```suggestion
</td>

```

File: wagtail/contrib/search_promotions/tests.py
Comment (laymonage) [inline wagtail/contrib/search_promotions/tests.py:899]: I think it'd be great to have another test case with a user that only has a minimal set of permissions, e.g. just the `edit` permission for the `Query` model to prove that it works with a non-superuser.

File: wagtail/contrib/search_promotions/views.py
Comment (laymonage) [inline wagtail/contrib/search_promotions/views.py:69]: It'd also be great to have a test case for the number of queries on the index page when we have `Query` objects with multiple `SearchPromotion`s, so we can clearly see how significant the improvement is.

**Review by laymonage — APPROVED** [review]
Thanks, this is really great work! Only one minor comment for the implementation, the rest is about adding more tests to ensure that all the previous points I raised are actually tested (and they will fail if implemented incorrectly). I've already made these changes locally so I'll push and merge right now.

I think this is a really good example of how we refactor a legacy function-based view to use the new generic views. It doesn't use a filterset though, so it's not a complete example, but we have other PRs like #11326 and #11304 for these examples.

Once again, thank you very much for this!

File: wagtail/contrib/search_promotions/tests.py
Comment (laymonage) [inline wagtail/contrib/search_promotions/tests.py]: !!! 🤩

**Review by laymonage — COMMENTED** [review]
I'll wait for the CI to pass, then merge.

File: wagtail/contrib/search_promotions/tests.py
Comment (zerolab) [reply wagtail/contrib/search_promotions/tests.py]: ❤️

**Comment by rohitsrma** [comment]
Thanks @laymonage for all the amazing feedbacks. I really enjoyed working on this. Looking forward to trying my hand on the remaining FBVs! 🙌

**Pull request merged**
